src/extension.ts

```diff
@@ -1,45 +1,142 @@
+                        enabledOrNull(() =>
+                            observableOrPromiseCompat(h.definition(doc, pos))
+                        ),
                 }
-                await config.update('extensions', extensions)
+            )
+        )
+    )
+    ctx.subscriptions.add(
+        reregisterWhenEnablementChanges(() =>
+            sourcegraph.languages.registerReferenceProvider(DOCUMENT_SELECTOR, {
+                provideReferences: (doc, pos) =>
+                    enabledOrNull(() =>
+                        observableOrPromiseCompat(h.references(doc, pos))
+                    ),
             })
-            .catch(err => console.error(err))
-    })
-
-    sourcegraph.languages.registerDefinitionProvider(['*'], {
-        provideDefinition: (doc, pos) => h.definition(doc, pos),
-    })
-    sourcegraph.languages.registerReferenceProvider(['*'], {
-        provideReferences: (doc, pos) => h.references(doc, pos),
-    })
+        )
+    )
+}
+
+const settingsSubscribable = new Observable<Settings>(sub => {
+    sub.next(sourcegraph.configuration.get().value)
+    return sourcegraph.configuration.subscribe(() =>
+        sub.next(sourcegraph.configuration.get().value)
+    )
+})
+
+function enabledOrNull<T>(provider: () => T): T | null {
+    if (
+        !sourcegraph.configuration.get<Settings>().value[
+            'basicCodeIntel.enabled'
+        ]
+    ) {
+        return null
+    }
+    return provider()
+}
+
+/**
+ * This makes it so that basicCodeIntel.toggle (the "Show/hide fuzzy matches" button) immediately takes effect and
+ * changes the locations that are currently being displayed in the panel.
+ *
+ * If we used an observable instead, it would always show the loading indicator.
+ */
+function reregisterWhenEnablementChanges(
+    register: () => sourcegraph.Unsubscribable
+): sourcegraph.Unsubscribable {
+    let registration: sourcegraph.Unsubscribable | undefined
+    return from(settingsSubscribable)
+        .pipe(
+            distinctUntilChanged(
+                (a, b) =>
+                    Boolean(a['basicCodeIntel.enabled']) ===
+                        Boolean(b['basicCodeIntel.enabled']) &&
+                    a['basicCodeIntel.definition.symbols'] ===
+                        b['basicCodeIntel.definition.symbols']
+            ),
+            map(() => {
+                if (registration) {
+                    registration.unsubscribe()
+                }
+                registration = register()
+            }),
+            finalize(() => {
+                if (registration) {
+                    registration.unsubscribe()
+                    registration = undefined
+                }
+            })
+        )
+        .subscribe()
+}
+
+function observableOrPromiseCompat<T>(
+    result: Observable<T> | Promise<T>
+): sourcegraph.ProviderResult<T> {
+    // HACK: Earlier extension API versions did not support providers returning observables. We can detect whether
```

Comment:
We should plan to handle the issue of extension API breakage differently - IMO there should be no backwards compatibility handling in extension code, it should be handled at the registry/extension host level.

eg.:
- If the user has activated the Sourcegraph extension E, which is only compatible with extension API vN, and his browser extension packages extension API vN+1, he receives a notification that "Incompatible extension E could not be activated"
- If the user's browser extension packages extension API vN, the extension registry will provide the latest version of E that was compatible with extension API vN, even if the very latest version of E on the registry was coded against vN+1

Reply:
:+1: Moved to https://github.com/sourcegraph/sourcegraph/issues/1255